bigtable-hbase-1.x-parent/bigtable-hbase-1.x-mapreduce/src/main/java/com/google/cloud/bigtable/mapreduce/hbasesnapshots/ImportHBaseSnapshotJob.java

```diff
@@ -0,0 +1,290 @@
+            + " to "
+            + conf.get(TABLENAME_KEY));
+
+    // implicit bigtable configs
+    BigtableConfiguration.configure(
+        conf,
+        conf.get(BigtableOptionsFactory.PROJECT_ID_KEY),
+        conf.get(BigtableOptionsFactory.INSTANCE_ID_KEY),
+        conf.get(BigtableOptionsFactory.APP_PROFILE_ID_KEY, ""));
+    conf.setBooleanIfUnset(BigtableOptionsFactory.BIGTABLE_USE_TIMEOUTS_KEY, true);
+    conf.setIfUnset(
+        BigtableOptionsFactory.BIGTABLE_MUTATE_RPC_TIMEOUT_MS_KEY,
+        String.valueOf(BIGTABLE_MUTATE_RPC_TIMEOUT_MS_DEFAULT));
+
+    // implicit table outputformat configs that are used in the job to write map output to a table
+    conf.set(TableOutputFormat.OUTPUT_TABLE, conf.get(TABLENAME_KEY));
+    conf.setStrings(
+        "io.serializations",
+        conf.get("io.serializations"),
+        MutationSerialization.class.getName(),
+        ResultSerialization.class.getName());
+
+    return 0;
+  }
+
+  /**
+   * Sets up the actual job.
+   *
+   * @param conf The current configuration.
+   * @return The newly created job.
+   * @throws java.io.IOException When setting up the job fails.
+   */
+  protected static Job createSubmittableJob(Configuration conf) throws IOException {
+    Job job = Job.getInstance(conf, conf.get(IMPORT_SNAPSHOT_JOBNAME_KEY));
+    job.setJarByClass(ImportHBaseSnapshotJob.class);
+    TableMapReduceUtil.initTableSnapshotMapperJob(
+        conf.get(SNAPSHOTNAME_KEY),
+        new Scan().setMaxVersions(),
+        SnapshotMapper.class,
+        ImmutableBytesWritable.class,
+        Writable.class,
+        job,
+        true,
+        new Path(conf.get(SNAPSHOT_RESTOREDIR_KEY)));
+
+    job.setNumReduceTasks(0);
+    job.setOutputFormatClass(TableOutputFormat.class);
+
+    return job;
+  }
+
+  protected enum ScanCounter {
+    NUM_ROWS,
+    NUM_CELLS,
+  }
+
+  /** MAPPER */
+  protected static class SnapshotMapper extends TableMapper<ImmutableBytesWritable, Put> {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(SnapshotMapper.class);
+
+    private Boolean isEmptyRowWarned = false;
+
+    protected static final int MAX_CELLS = 100_000 - 1;
+
+    @Override
+    protected void setup(Context ctx) {
+      // noop
+    }
+
+    @Override
+    protected void map(ImmutableBytesWritable key, Result result, Context ctx)
+        throws IOException, InterruptedException {
+
+      List<Cell> cells = checkEmptyRow(result);
+      if (cells.isEmpty()) {
+        return;
+      }
+
+      ctx.getCounter(ScanCounter.NUM_ROWS).increment(1);
+      ctx.getCounter(ScanCounter.NUM_CELLS).increment(cells.size());
+
+      // Split the row into multiple puts if it exceeds the maximum mutation limit
+      Iterator<Cell> cellIt = cells.iterator();
+
+      while (cellIt.hasNext()) {
+        Put put = new Put(key.get());
+
+        for (int i = 0; i < MAX_CELLS && cellIt.hasNext(); i++) {
+          put.add(cellIt.next());
+        }
```

Comment:
This looks like silent data loss. I think it needs a bit more thought

Reply:
this mimics the behavior in sequence file import [HBaseResultToMutationFn.java#L100](https://github.com/googleapis/java-bigtable-hbase/blob/master/bigtable-dataflow-parent/bigtable-beam-import/src/main/java/com/google/cloud/bigtable/beam/sequencefiles/HBaseResultToMutationFn.java#L100) and we test this condition in [TestSnapshotMapper.testRowExceedingMaxCells()](https://github.com/jhambleton/java-bigtable-hbase/blob/master/bigtable-hbase-1.x-parent/bigtable-hbase-1.x-mapreduce/src/test/java/com/google/cloud/bigtable/mapreduce/hbasesnapshots/TestSnapshotMapper.java#L111) and assert if 2 rows are emitted and on number of cells matching. 

i've reviewed this again and don't see conditions where the iterator would escape the loop and silently lose data. lmk if you had anything in mind and i'd be happy to incorporate it

Reply:
Sorry should've been more clear. If there are more cells than MAX_CELLS then they will be silently dropped.

Reply:
my mistake